Fall back to copying when os.link raises error

chmreid opened the pull request:
This addresses a permission denied error on NFS file systems that is caused by the `os.link` call. Addresses #519 

Previously we were catching a "too many links" exception, now we are also catching a "permission denied" exception and an "operation not permitted" exception. (Note that we may be able to remove the "operation not permitted" exception number from the list of exceptions that lead us to fall back to copying.)

### Useful Links

Useful discussion of possible approaches to solving this better: https://bugs.python.org/issue37612

Useful info about NFS and Python's os module: https://web.archive.org/web/20100912144722/http://www.unixcoding.org/NFSCoding

Note: this PR does not (yet) have tests for the new error codes being caught/handled. Tests TBA.

jessebrennan commented:
@chmreid I see that you requested my review, but didn't assign the PR to me. I'm confused by this. Did you see this Slack thread? https://ucsc-cgl.slack.com/archives/C705Y6G9Z/p1585093339100200

chmreid commented:
@jessebrennan The problem is with the `master` branch, nothing specific to this PR. I tried re-running Travis tests for master and they failed: https://travis-ci.org/github/HumanCellAtlas/dcp-cli/builds/654957449

We can pick up on this PR once master is passing.

chmreid commented:
@jessebrennan Tests are passing, can you give this a review now?

hannes-ucsc commented:
I would like the opportunity to review this. I'm currently not assigned so I'll hold off for now.

chmreid commented:
I have assigned this issue to you, please review when you get the chance. Thanks!

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -58,7 +58,13 @@ def hardlink(source, link_name):
             raise
     except OSError as e:
-        if e.errno == errno.EMLINK:
+        # See https://docs.python.org/3/library/errno.html
```
```suggestion
```

Redundant, any Python programmer can locate the documentation for the standard library. We don't link to `logging`, `shutil` etc. and `errno` is no different.

chmreid replied:
Comment removed

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -58,7 +58,13 @@ def hardlink(source, link_name):
-        if e.errno == errno.EMLINK:
+        # See https://docs.python.org/3/library/errno.html
+        copy_on_error = [
+            errno.EMLINK,  # too many links
+            errno.EACCES,  # permission denied
+            errno.EPERM,  # operation not permitted
```
```suggestion
            errno.EMLINK,
            errno.EACCES,
            errno.EPERM,
```

The meaning of the constants is [well documented](https://docs.python.org/3.6/library/errno.html#errno.EMLINK). We don't need to duplicate that here. What isn't obvious is the intent e.g., *why* EPERM is included. Like that we expect (and have observed) this to occur on an NFS share.

chmreid replied:
Comments removed, better descriptive (why EPERM or EMLINK) added

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -58,7 +58,13 @@ def hardlink(source, link_name):
     except OSError as e:
-        if e.errno == errno.EMLINK:
+        # See https://docs.python.org/3/library/errno.html
+        copy_on_error = [
```
We typically use tuples for immutable sequences that are used with `in`. Not for performance reasons but to more clearly indicate intent.

chmreid replied:
Changed to tuple

hannes-ucsc commented on hca/dss/util/__init__.py:
```diff
@@ -58,7 +58,13 @@ def hardlink(source, link_name):
+        # See https://docs.python.org/3/library/errno.html
+        copy_on_error = [
+            errno.EMLINK,  # too many links
+            errno.EACCES,  # permission denied
```
```suggestion
```

Why is EACCES included here? Our current philosophy is to catch only expected errors that we know can be addressed by falling back to a copy. I have seen no evidence that EACCES is to be expected and should be handled by falling back to a copy. Mere similarity isn't sufficient evidence for me. In fact, I suspect that it is very likely that if `os.link` fails with EACCES, so will `open(…, 'w')` which is what `shutil.copyfile` uses under the hood.

chmreid replied:
This is a good point, removed EACCES so now it is just EMLINK and EPERM

chmreid commented:
@hannes-ucsc Thank you for the comments. I have addressed your feedback, and this PR is ready for another review. Would you please review it at your convenience? Thanks.
